packages/aws-cdk/lib/api/hotswap-deployments.ts

```diff
@@ -57,24 +58,81 @@ export async function tryHotswapDeployment(
+    } else if (resourceHotswapEvaluation === ChangeHotswapImpact.IRRELEVANT) {
+      // empty 'if' just for flow-aware typing to kick in...
+    } else {
+      const detectorResults = new Array<Promise<ChangeHotswapResult>>();
+      detectorResults.push(isHotswappableLambdaFunctionChange(logicalId, resourceHotswapEvaluation, evaluateCfnTemplate));
+      detectorResults.push(isHotswappableStateMachineChange(logicalId, resourceHotswapEvaluation, evaluateCfnTemplate));
+
+      promises.push(detectorResults);
+    }
   });
-  return Promise.all(promises).then(hotswapDetectionResults => {
-    const hotswappableResources = new Array<HotswapOperation>();
-    let foundNonHotswappableChange = false;
-    for (const lambdaFunctionShortCircuitChange of hotswapDetectionResults) {
-      if (lambdaFunctionShortCircuitChange === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT) {
+
+  const hotswappableResources = new Array<HotswapOperation>();
+  const changesDetectionResults: Array<Array<ChangeHotswapResult>> = [];
+
+  for (const detectorResultPromises of promises) {
+    const hotswapDetectionResults = await Promise.all(detectorResultPromises);
+    changesDetectionResults.push(hotswapDetectionResults);
+  }
+
+  // resolve all detector results
+  for (const hotswapDetectionResults of changesDetectionResults) {
+    const perChangeHotswappableResources = new Array<HotswapOperation>();
+
+    for (const result of hotswapDetectionResults) {
+      if (typeof result !== 'string') {
+        perChangeHotswappableResources.push(result);
+      }
+    }
+
+    // if we found any hotswappable changes, return now
+    if (perChangeHotswappableResources.length > 0) {
+      hotswappableResources.push(...perChangeHotswappableResources);
+      continue;
+    }
+
+    // no hotswappable changes found, so any REQUIRES_FULL_DEPLOYMENTs imply a non-hotswappable change
+    for (const result of hotswapDetectionResults) {
+      if (result === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT) {
         foundNonHotswappableChange = true;
-      } else if (lambdaFunctionShortCircuitChange === ChangeHotswapImpact.IRRELEVANT) {
-        // empty 'if' just for flow-aware typing to kick in...
-      } else {
-        hotswappableResources.push(lambdaFunctionShortCircuitChange);
       }
     }
-    return foundNonHotswappableChange ? undefined : hotswappableResources;
-  });
+    // no REQUIRES_FULL_DEPLOYMENT implies that all results are IRRELEVANT
+  }
+
+  return foundNonHotswappableChange ? undefined : hotswappableResources;
+}
+
+/**
+ * returns `ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT` if a resource was deleted, or a change that we cannot short-circuit occured.
+ * Returns `ChangeHotswapImpact.IRRELEVANT` if a change that does not impact shortcircuiting occured, such as a metadata change.
+ */
+export function isResourceChangeHotswappable(change: cfn_diff.ResourceDifference): HotswappableResourceChange | ChangeHotswapImpact {
```

Comment:
This name could use some word-smithing I think.

How about something like `isCandidateForHotswapping()`?